Backport better ES startup checks for test040CreateKeystoreManually

williamrandolph opened the pull request:
This test got moved to `KeystoreManagementTests` and significantly restructured on the 7.x and master branches. It doesn't seem to be causing trouble there. Back here on 7.6, it's been failing unpredictably - see #51316 for details.

I looked at the difference between what's happening in 7.6 and what's happening on master, and the difference seems to be more waiting for files to be available, especially after we create a keystore: https://github.com/elastic/elasticsearch/blob/master/qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java#L391

This PR brings 7.6 into line with the other branches, despite the different structure of the tests.

williamrandolph committed:
Add more waits in flaky Docker test

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

pugnascotia reviewed (approved):
LGTM. On first inspection, it seems odd that we'd need to wait for these files to exist if the preceding commands were successful, but I can also believe there's enough overlay magic going on to make the waits necessary.

williamrandolph commented:
@pugnascotia It is weird, isn't it? I'm going to spin up a debian 9 test environment and see if I can reproduce the original issue.

williamrandolph commented:
I let a Debian 9 GCP instance spin for a while running `distructiveDistroTest.default-docker` against the 7.6 branch, but I got no errors. Hmm.

williamrandolph committed:
Remove waits and add reliable check for ES startup

Tossing in wait statements to try to make things work is a code smell
and I shouldn't have done it.

Instead, I dug a little deeper and found a potential race condition that
I fixed in the course of the password-protected keystore work. The
Docker tests had been assuming that ES had started up if the ES bundled
JDK was running. But that JDK gets called in our startup scripts for
several other uses. It's better to get long output from ps and check for
the invocation of the Elasticsearch startup class.

williamrandolph commented:
I've taken another look at this, and I'm not at all convinced by my knee-jerk, code-smelly analysis. I'm going to close and take another shot at this.
